File: openharmony/arkcompiler_runtime_core/static_core/plugins/ets/tests/interop_js/tests/constructor/ts_to_sts/constructor.js
'use strict';
/**
 * Copyright (c) 2024 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the 'License');
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an 'AS IS' BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
let __extends = (this && this.__extends) || (function () {
    let extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (let p in b) { if (Object.prototype.hasOwnProperty.call(b, p)) { d[p] = b[p]; } } };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== 'function' && b !== null) {
            throw new TypeError('Class extends value ' + String(b) + ' is not a constructor or null'); }
        extendStatics(d, b);
        function FuC() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (FuC.prototype = b.prototype, new FuC());
    };
})();
Object.defineProperty(exports, '__esModule', { value: true });
exports.childClassInstance = exports.ChildClass = exports.abstractClassInstance = exports.AbstractClass = 
exports.simpleArrowFunction = exports.SimpleObject = exports.methodCreateClassInstance = 
exports.methodCreateAnonymousClass = exports.methodConstructorInstance = exports.MethodCreateConstructor = 
exports.IIFEConstructorInstance = exports.IIFEConstructor = exports.IIFEClassInstance = exports.IIFEClass = 
exports.functionConstructorInstance = exports.anonymousClassInstance = exports.AnonymousClass = exports.namedClassInstance = 
exports.NamedClass = exports.tsInt = void 0;
exports.createNamedClassFromTs = createNamedClassFromTs;
exports.createAnonymousClassFromTs = createAnonymousClassFromTs;
exports.FunctionConstructor = FunctionConstructor;
exports.createFunctionConstructorFromTs = createFunctionConstructorFromTs;
exports.createIIFEClassFromTs = createIIFEClassFromTs;
exports.createIIFEConstructorFromTs = createIIFEConstructorFromTs;
exports.createMethodConstructorClass = createMethodConstructorClass;
exports.createAbstractClassFromTs = createAbstractClassFromTs;
exports.createChildClassFromTs = createChildClassFromTs;
exports.tsInt = 1;
let NamedClass = /** @class */ (function () {
    function NamedClass(value) {
        this._value = value;
    }
    return NamedClass;
}());
exports.NamedClass = NamedClass;
function createNamedClassFromTs() {
    return new NamedClass(exports.tsInt);
}
exports.namedClassInstance = new NamedClass(exports.tsInt);
exports.AnonymousClass = /** @class */ (function () {
    function class1(value) {
        this.value = value;
        this._value = value;
    }
    return class1;
}());
function createAnonymousClassFromTs() {
    return new exports.AnonymousClass(exports.tsInt);
}
exports.anonymousClassInstance = new exports.AnonymousClass(exports.tsInt);
function FunctionConstructor(value) {
    this._value = value;
}
function createFunctionConstructorFromTs() {
    return new FunctionConstructor(exports.tsInt);
}
exports.functionConstructorInstance = new FunctionConstructor(exports.tsInt);
exports.IIFEClass = (function () {
    return /** @class */ (function () {
        function class2(value) {
            this._value = value;
        }
        return class2;
    }());
})();
function createIIFEClassFromTs() {
    return new exports.IIFEClass(exports.tsInt);
}
exports.IIFEClassInstance = new exports.IIFEClass(exports.tsInt);
exports.IIFEConstructor = (function () {
    function constructorFoo(value) {
        this._value = value;
    }
    return constructorFoo;
})();
function createIIFEConstructorFromTs() {
    return new exports.IIFEConstructor(exports.tsInt);
}
exports.IIFEConstructorInstance = new exports.IIFEConstructor(exports.tsInt);
let MethodCreateConstructor = /** @class */ (function () {
    function MethodCreateConstructor() {
    }
    MethodCreateConstructor.prototype.Constructor = function () {
        return /** @class */ (function () {
            function class3(value) {
                this._value = value;
            }
            return class3;
        }());
    };
    return MethodCreateConstructor;
}());
exports.MethodCreateConstructor = MethodCreateConstructor;
function createMethodConstructorClass() {
    return new MethodCreateConstructor();
}
exports.methodConstructorInstance = new MethodCreateConstructor();
exports.methodCreateAnonymousClass = exports.methodConstructorInstance.Constructor();
exports.methodCreateClassInstance = new exports.methodCreateAnonymousClass(exports.tsInt);
exports.SimpleObject = {
    _value: exports.tsInt,
};
let simpleArrowFunction = function (int) {
    return {
        int: int,
    };
};
exports.simpleArrowFunction = simpleArrowFunction;
let Abstract = /** @class */ (function () {
    function abstract(value) {
        this._value = value;
    }
    return abstract;
}());
let AbstractClass = /** @class */ (function (_super) {
    __extends(AbstractClass, _super);
    function AbstractClass(...arg) {
        return _super !== null && _super.apply(this, arg) || this;
    }
    return AbstractClass;
}(Abstract));
exports.AbstractClass = AbstractClass;
function createAbstractClassFromTs() {
    return new AbstractClass(exports.tsInt);
}
exports.abstractClassInstance = new AbstractClass(exports.tsInt);
let ChildClass = /** @class */ (function (_super) {
    __extends(ChildClass, _super);
    function ChildClass(value) {
        return _super.call(this, value) || this;
    }
    return ChildClass;
}(NamedClass));
exports.ChildClass = ChildClass;
function createChildClassFromTs() {
    return new ChildClass(exports.tsInt);
}
exports.childClassInstance = new ChildClass(exports.tsInt);

File: openharmony/arkcompiler_runtime_core/static_core/plugins/ets/tests/interop_js/tests/constructor/ts_to_sts/constructor.cpp
/**
 * Copyright (c) 2024 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <gtest/gtest.h>
#include "ets_interop_js_gtest.h"

namespace ark::ets::interop::js::testing {

class EtsConstructorTsToEtsTest : public EtsInteropTest {};

TEST_F(EtsConstructorTsToEtsTest, checkNamedClass)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkNamedClass"));
}

TEST_F(EtsConstructorTsToEtsTest, checkCreateNamedClassFromTs)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkCreateNamedClassFromTs"));
}

TEST_F(EtsConstructorTsToEtsTest, checkNamedClassInstance)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkNamedClassInstance"));
}

TEST_F(EtsConstructorTsToEtsTest, checkAnonymousClass)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkAnonymousClass"));
}

TEST_F(EtsConstructorTsToEtsTest, checkCreateAnonymousClassFromTs)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkCreateAnonymousClassFromTs"));
}

TEST_F(EtsConstructorTsToEtsTest, checkAnonymousClassInstance)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkAnonymousClassInstance"));
}

TEST_F(EtsConstructorTsToEtsTest, checkFunctionConstructor)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkFunctionConstructor"));
}

TEST_F(EtsConstructorTsToEtsTest, checkCreateFunctionConstructorFromTs)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkCreateFunctionConstructorFromTs"));
}

TEST_F(EtsConstructorTsToEtsTest, checkFunctionConstructorInstance)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkFunctionConstructorInstance"));
}

TEST_F(EtsConstructorTsToEtsTest, checkIIFEClass)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkIIFEClass"));
}
// NOTE issue(18074) fix IIFE without keyword 'new'.
TEST_F(EtsConstructorTsToEtsTest, DISABLED_checkIIFEClassError)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkIIFEClassError"));
}

TEST_F(EtsConstructorTsToEtsTest, checkCreateIIFEClassFromTs)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkCreateIIFEClassFromTs"));
}

TEST_F(EtsConstructorTsToEtsTest, checkIIFEClassInstance)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkIIFEClassInstance"));
}

TEST_F(EtsConstructorTsToEtsTest, checkIIFEConstructor)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkIIFEConstructor"));
}

TEST_F(EtsConstructorTsToEtsTest, checkIIFEConstructorUndefined)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkIIFEConstructorUndefined"));
}

TEST_F(EtsConstructorTsToEtsTest, checkCreateIIFEConstructorFromTs)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkCreateIIFEConstructorFromTs"));
}

TEST_F(EtsConstructorTsToEtsTest, checkIIFEConstructorInstance)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkIIFEConstructorInstance"));
}
// NOTE issue(18077) fix try to call method with constructor.
TEST_F(EtsConstructorTsToEtsTest, DISABLED_checkMethodCreateConstructor)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkMethodCreateConstructor"));
}
// NOTE issue(18077) fix try to call method with constructor.
TEST_F(EtsConstructorTsToEtsTest, DISABLED_checkCreateMethodConstructorClass)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkCreateMethodConstructorClass"));
}
// NOTE issue(18077) fix try to call method with constructor.
TEST_F(EtsConstructorTsToEtsTest, DISABLED_checkMethodConstructorInstance)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkMethodConstructorInstance"));
}

TEST_F(EtsConstructorTsToEtsTest, checkMethodCreateAnonymousClass)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkMethodCreateAnonymousClass"));
}

TEST_F(EtsConstructorTsToEtsTest, checkMethodCreateClassInstance)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkMethodCreateClassInstance"));
}

TEST_F(EtsConstructorTsToEtsTest, checkSimpleObject)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkSimpleObject"));
}
// NOTE issue(18072) fix create object with keyword new .
TEST_F(EtsConstructorTsToEtsTest, DISABLED_checkSimpleArrowFunction)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkSimpleArrowFunction"));
}

TEST_F(EtsConstructorTsToEtsTest, checkAbstractClass)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkAbstractClass"));
}

TEST_F(EtsConstructorTsToEtsTest, checkCreateAbstractClassFromTs)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkCreateAbstractClassFromTs"));
}

TEST_F(EtsConstructorTsToEtsTest, checkAbstractInstance)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkAbstractInstance"));
}

TEST_F(EtsConstructorTsToEtsTest, checkChildClass)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkChildClass"));
}

TEST_F(EtsConstructorTsToEtsTest, checkCreateChildClassFromTs)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkCreateChildClassFromTs"));
}

TEST_F(EtsConstructorTsToEtsTest, checkChildInstance)
{
    ASSERT_EQ(true, CallEtsMethod<bool>("checkChildInstance"));
}

}  // namespace ark::ets::interop::js::testing
